File: snippets/java/src/main/java/services/CancellationExample.java
package services;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.common.TerminalException;
import foundations.actions.NotificationService;
import usecases.microservices.utils.Order;

@VirtualObject
public class CancellationExample {

  StateKey<String> STATUS = StateKey.of("status", String.class);

  // <start_here>
  @Handler
  public void processOrder(Order order) {
    // If cancellation happened before this line, this still executes
    Restate.state().set(STATUS, "processing");

    // If cancellation happened before this line, this still executes
    String paymentId = Restate.random().nextUUID().toString();
    try {

      // If canceled right before this line, Restate.run won't execute
      // If canceled during run block execution,
      // then a terminal exception gets thrown here once execution finishes
      Restate.run("process-payment", () -> processPayment(paymentId, order));

      // If cancellation happened right before this line, this still executes
      Restate.serviceHandle(NotificationService.class)
          .send(NotificationService::sendNotification, "Your order is processed");

    } catch (TerminalException e) {
      // Cancellation detected - run compensation
      Restate.run("refund-payment", () -> refundPayment(paymentId, order));
      throw e; // Re-throw to propagate cancellation
    }
  }

  // <end_here>

  private void processPayment(String paymentId, Order order) {}

  private void refundPayment(String paymentId, Order order) {}
}

File: snippets/ts/src/services/cancellation-example.ts
import * as restate from "@restatedev/restate-sdk";

// Type definitions
interface Order {
  userId: string;
  amount: number;
  id: string;
}

// Service stubs
const notificationService = restate.service({
  name: "notificationService",
  handlers: {
    notify: async (ctx: restate.Context, message: string) => {},
  },
});
// Function stubs
async function processPayment(
  paymentId: string,
  order: Order
): Promise<string> {
  // Simulate payment processing
  return "payment_id_" + order.id;
}

async function refundPayment(paymentId: string, order: Order): Promise<void> {
  // Simulate refund processing
}

// <start_here>
async function processOrder(ctx: restate.ObjectContext, order: Order) {
  // If cancellation happened before this line, this still executes
  ctx.set("status", "processing");

  // If cancellation happened before this line, this still executes
  const paymentId = ctx.rand.uuidv4().toString();

  try {
    // If cancelled right before this line, ctx.run won't execute
    // If cancelled during run block execution,
    // then a terminal error gets thrown here once execution finishes
    await ctx.run(() => processPayment(paymentId, order));

    // If cancellation happened before this line, this still executes
    ctx.serviceSendClient(notificationService).notify("Payment processed");
  } catch (error) {
    if (error instanceof restate.TerminalError) {
      // Cancellation detected - run compensation
      await ctx.run(() => refundPayment(paymentId, order));
      throw error; // Re-throw to propagate cancellation
    }
  }
}
// <end_here>
